### ISSUE TITLE
plugin failure, but exit code 0

### ISSUE THREAD

**Issue by mcast** [wtsi-npg/baton#218]
I haven't finished investigating yet, but the basic problem is the `baton-metamod` returns exit code zero (so I assume it was successful) but it failed.

```
ubuntu@mca-icli427:~$ i-cgp -v /software/sciops/pkgg/baton/2.0.1/bin/baton-metamod --operation add --file /tmp/batonq.json.pElT_5
[d] IRODS_ENVIRONMENT_FILE=/home/ubuntu/.irods/irods_environment-cgp.json
[d] exec /software/sciops/pkgg/baton/2.0.1/bin/baton-metamod
ERROR: [-]      iRODS/lib/core/src/sockComm.cpp:1288:sendStartupPack :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message []
        [-]     iRODS/lib/core/src/sockComm.cpp:1364:sendRodsMsg :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message [failed to resolve network interface]
                [-]     iRODS/lib/core/src/irods_tcp_object.cpp:95:resolve :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message []
                        [-]     iRODS/lib/core/src/irods_network_manager.cpp:93:init_from_type :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message [Failed to load network plugin]
                                [-]     iRODS/lib/core/src/irods_network_manager.cpp:74:load_network_plugin :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message []
                                        [-]     iRODS/lib/core/include/irods_load_plugin.hpp:155:load_plugin :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message []
                                                [-]     iRODS/lib/core/include/irods_load_plugin.hpp:53:resolve_plugin_path :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message [does not exist [/var/lib/irods/plugins/network]]

ERROR: connectToRhost: sendStartupPack to irods-cgp-g1-vm.internal.sanger.ac.uk failed status = -130000 SYS_INVALID_INPUT_PARAM
ERROR: _rcConnect: connectToRhost error, server on irods-cgp-g1-vm.internal.sanger.ac.uk:1247 is probably down status = -130000 SYS_INVALID_INPUT_PARAM
2020-03-27T20:18:43Z ERROR Failed to connect to irods-cgp-g1-vm.internal.sanger.ac.uk:1247 zone 'cgp' as 'mca'
ubuntu@mca-icli427:~$ echo $?
0
```

The absence of `/var/lib/irods/plugins/network` is real and caused by my choice of `irods-icommands` package, and is a change from my icommands 4.1.12 to 4.2.7 which I've just discovered.

Should I check stderr before I assume success?  
Or is it possible for this irods init failure to be an error, and then I can require a minimum baton version?

**Comment by kjsanger** [comment]
Thanks for reporting. That's a bug. `error_count` should be incremented here:

https://github.com/wtsi-npg/baton/blob/devel/src/operations.c#L59

to trigger a non-zero exit. Fix confirmed locally and will be in the next bugfix release (`2.0.2`) which should be ready in April.

**Comment by mcast** [comment]
Thanks...  I'm curious because https://github.com/wtsi-npg/baton/blob/2.0.1/src/baton.c#L146 error message is triggered. The only other place I see that text is #161 where it looks like exit code 5, so some regression?  Something hooking `atexit`?

The history of the "install" of the binary is dubious and I should declare it, in case it's relevant.
1. I used the (internal / company local) ISG provided Openstack image with icommands 4.1.12 and copied `/software/sciops/pkgg/baton/2.0.1/` from an internal bionic machine. Lazy but quick and it worked fine.
2. I started a fresh vm with the (local) `bionic-WTSI-irodsclient_427_8ff4b8` image and repeated, but the `irods-icommands` package doesn't have...

Ah, I forgot to check dependency packages! I found the plugins under `/usr/lib/irods/plugins/` ...which is a more LSB-ish place than `/var/lib/irods/plugins/network` anyway.

**Comment by kjsanger** [comment]
You have a point there. Incrementing the error count is an easy "fix", but doesn't really keep the correct behaviour wrt counting errors per JSON document.

**Comment by kjsanger** [comment]
I've made a change that will separate the error count from the success/fail of the function so that both can be captured.

**Comment by mcast** [comment]
I now have a working `baton-metamod` and demonstrated that feeding this an invalid path through `.irods/irods_environment.json "irods_plugins_home": ...` brings on the same error.

**Comment by kjsanger** [comment]
The same error in a good or bad way?

**Comment by mcast** [comment]
In a consistent way, when provoked by a deliberate configuration error instead of a Gonzo the Muppet build method 👹 , so hopefully that makes it easier to confirm fixed & auto-test.

**Comment by kjsanger** [comment]
So you get a non-zero exit and it says it processed 0 items with 0 errors? (Which is what we want.)

**Comment by mcast** [comment]
Ignoring the distraction of a new(?) problem #220.

No, with
* https://github.com/wtsi-npg/baton/releases/download/2.0.1/baton-2.0.1.tar.gz which I built with https://packages.irods.org/apt/pool/xenial/main/i/irods-dev/irods-dev_4.2.7_amd64.deb
* when correctly configured, it works OK
* but I deliberately configured a missing plugin
then it fails in the mode of this bug #218.

```
ubuntu@mca-icli427:~$ i-cgp -v /usr/local/bin/baton-metamod --operation add --file ~/batonq.json 
[d] IRODS_ENVIRONMENT_FILE=/home/ubuntu/.irods/irods_environment-cgp.json
[d] exec /usr/local/bin/baton-metamod
remote addresses: 172.30.130.241 ERROR: [-]     /irods/lib/core/src/sockComm.cpp:1159:int sendStartupPack(rcComm_t *, int, int) :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message [failed to resolve network interface]
        [-]     /irods/lib/core/src/sockComm.cpp:1235:irods::error sendRodsMsg(irods::network_object_ptr, const char *, bytesBuf_t *, bytesBuf_t *, bytesBuf_t *, int, irodsProt_t) :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message [failed to resolve network interface]
                [-]     /irods/lib/core/src/irods_tcp_object.cpp:98:virtual irods::error irods::tcp_object::resolve(const std::string &, plugin_ptr &) :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message [Failed to load network plugin]
                        [-]     /irods/lib/core/src/irods_network_manager.cpp:104:irods::error irods::network_manager::init_from_type(const int &, const std::string &, const std::string &, const std::string &, const std::string &, network_ptr &) :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message [Failed to load network plugin]
                                [-]     /irods/lib/core/src/irods_network_manager.cpp:75:irods::error irods::load_network_plugin(network_ptr &, const std::string &, const std::string &, const std::string &) :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message [does not exist [/absent/irods/plugins/network]]
                                        [-]     /irods/lib/core/include/irods_load_plugin.hpp:137:irods::error irods::load_plugin(PluginType *&, const std::string &, const std::string &, const std::string &, const Ts &...) [PluginType = irods::network, Ts = <std::__1::basic_string<char>>] :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message [does not exist [/absent/irods/plugins/network]]
                                                [-]     /irods/lib/core/include/irods_load_plugin.hpp:66:irods::error irods::resolve_plugin_path(const std::string &, std::string &) :  status [SYS_INVALID_INPUT_PARAM]  errno [] -- message [does not exist [/absent/irods/plugins/network]]

remote addresses: 172.30.130.241 ERROR: connectToRhost: sendStartupPack to irods-cgp-g1-vm.internal.sanger.ac.uk failed status = -130000 SYS_INVALID_INPUT_PARAM
 ERROR: _rcConnect: connectToRhost error, server on irods-cgp-g1-vm.internal.sanger.ac.uk:1247 is probably down status = -130000 SYS_INVALID_INPUT_PARAM
2020-03-31T11:20:15Z ERROR Failed to connect to irods-cgp-g1-vm.internal.sanger.ac.uk:1247 zone 'cgp' as 'mca'
ubuntu@mca-icli427:~$ echo $?
0
ubuntu@mca-icli427:~$ grep absent ~/.irods/irods_environment-cgp.json
  "irods_plugins_home": "/absent/irods/plugins/",
```

It doesn't tell how many items or errors.

**Comment by kjsanger** [comment]
Sorry, I assumed that you'd built from devel with the fix to test that it now did what you expect. That behaviour with `2.0.1` is expected as it it's the same version you reported this bug against.

**Comment by kjsanger** [comment]
Fixed by 4525291329d0b89c5be6b5b6f3bcd29d68c659bd
